File: java/src/peachtree/epi/EpiAPI.java
package peachtree.epi;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import org.json.JSONObject;

import peachtree.aln.Alignment;
import peachtree.aln.AlignmentAPI;
import peachtree.options.OptionsAPI;
import peachtree.phy.PhylogenyAPI;
import peachtree.phy.Tree;

public class EpiAPI {
	
	
	private static Epidemiology EPIDEMIOLOGY = null;
	
	private static boolean epiAccessionsAreDirty = false;
	private static boolean epiAnnotationsAreDirty = false;
	
	
	/**
	 * Upload epidmiological information as a csv/tsv
	 * @param contents
	 * @return
	 */
	public static String uploadEpi(String contents) {
		
		System.out.println("Uploading epidemiological information");
		
		
		try {
			long start = Calendar.getInstance().getTimeInMillis();
			
			
			EPIDEMIOLOGY = new Epidemiology();
			EPIDEMIOLOGY.parseFile(contents);
			EpiAPI.setEpiAccessionsToDirty();
			EpiAPI.setEpiAnnotationsToDirty();
			EpiAPI.validateAccessions(AlignmentAPI.getAlignment());
			EpiAPI.addAnnotationsToTree(PhylogenyAPI.getTree());
			
			OptionsAPI.prepareEpiAnnotations();
			
			long finish = Calendar.getInstance().getTimeInMillis();
			JSONObject json = new JSONObject().put("time", (finish-start)/1000.0);
			return json.toString();
			
			
		}catch (Exception e){
			EPIDEMIOLOGY = null;
			e.printStackTrace();
			return OptionsAPI.getErrorJSON(e);
		}
		
		
	}
	
	
	/**
	 * Indicate that the accessions need to be revalidated
	 */
	public static void setEpiAccessionsToDirty() {
		epiAccessionsAreDirty = true;
	}
	
	
	/**
	 * Indicate that the annotations need to be added to the tree
	 */
	public static void setEpiAnnotationsToDirty() {
		epiAnnotationsAreDirty = true;
	}
	
	
	
	
	/**
	 * Validate the accessions in the epidemiology with those in the alignment
	 * @param alignment
	 * @throws Exception
	 */
	public static void validateAccessions(Alignment alignment) throws Exception {
		if (!epiAccessionsAreDirty) return;
		if (alignment == null) return;
		if (EPIDEMIOLOGY == null) return;
		EPIDEMIOLOGY.validateAccessions(alignment);
		epiAccessionsAreDirty = false;
	}
	
	
	
	
	/**
	 * Add annotations to the tree
	 * @param epi
	 * @throws Exception
	 */
	public static void addAnnotationsToTree(Tree tree) throws Exception {
		if (!epiAnnotationsAreDirty) return;
		if (tree == null) return;
		if (EPIDEMIOLOGY == null) return;
		EPIDEMIOLOGY.addAnnotationsToTree(tree);
		epiAnnotationsAreDirty = false;
	}
	


	public static List<String> getAllAnnotations() {
		if (EPIDEMIOLOGY == null) return new ArrayList<String>();
		return EPIDEMIOLOGY.getAnnotations();
	}

}

File: cpp/peachtree/src/api/EpiAPI.cpp
/*
 * EpiAPI.cpp
 *
 *  Created on: 14/07/2021
 *      Author: jdou557
 */


#include <chrono>

#include "EpiAPI.h"
#include "PhylogenyAPI.h"
#include "AlignmentAPI.h"
#include "OptionsAPI.h"
#include "../error/Error.h"

using namespace std::chrono;


Epidemiology* EpiAPI::EPIDEMIOLOGY = nullptr;
Timeline* EpiAPI::timeline = nullptr;
bool EpiAPI::epiAccessionsAreDirty = false;
bool EpiAPI::epiAnnotationsAreDirty = false;



void EpiAPI::setEpi(Epidemiology* epi){
	EPIDEMIOLOGY = epi;
	epiAccessionsAreDirty = true;
	epiAnnotationsAreDirty = true;
}


/*
 * Get timeline graphics
 */
jsonObject EpiAPI::getTimelineGraphics(Node* subtree, Scaling* scaling, double axisFontSize, string symptomDateVar,
				int infectiousDaysBefore, int infectiousDaysAfter, string isolationDateVar, double branchWidth){
	if (EpiAPI::timeline == nullptr || !EpiAPI::timeline->isReady()) {
		jsonObject arr = json::array();
		return arr;
	}
	return EpiAPI::timeline->getTimelineGraphics(subtree, scaling, axisFontSize, symptomDateVar, infectiousDaysBefore, infectiousDaysAfter,
			isolationDateVar, branchWidth);
}


bool EpiAPI::isReady(){
	return EpiAPI::EPIDEMIOLOGY != nullptr;
}

/*
 * Reset sample heights if the subtree has changed
 */
void EpiAPI::prepareNodeSampleHeights(Node* subtree, string symptomDateVar, int infectiousDaysBefore, int infectiousDaysAfter){
	if (EpiAPI::timeline == nullptr) return;
	EpiAPI::timeline->prepareNodeSampleHeights(subtree, symptomDateVar, infectiousDaysBefore, infectiousDaysAfter);
}


/*
 * Prepare timeline
 */
void EpiAPI::prepareTimeline(Tree* tree, string sampleDateVariable, string dateFormat){

	if (EpiAPI::EPIDEMIOLOGY == nullptr) return;
	if (timeline == nullptr || epiAccessionsAreDirty){
		if (timeline != nullptr) {
			timeline->cleanup();
			delete timeline;
		}
		timeline = new Timeline(tree, EpiAPI::EPIDEMIOLOGY, sampleDateVariable, dateFormat);
	}else{
		timeline->setSampleDateVariable(sampleDateVariable, dateFormat);
	}

}

/*
 * Indicate that the accessions need to be revalidated
 */
void EpiAPI::setEpiAccessionsToDirty(){
	epiAccessionsAreDirty = true;
}

/*
 * Return the timeline
 */
Timeline* EpiAPI::getTimeline(){
	return EpiAPI::timeline;
}

/*
 * Indicate that the annotations need to be added to the tree
 */
void EpiAPI::setEpiAnnotationsToDirty(){
	epiAnnotationsAreDirty = true;
}


/*
 * Validate the accessions in the epidemiology with those in the alignment
 */
void EpiAPI::validateAccessions(Alignment* alignment){
	if (!epiAccessionsAreDirty) return;
	if (alignment == nullptr) return;
	if (EPIDEMIOLOGY == nullptr) return;
	EPIDEMIOLOGY->validateAccessions(alignment);
	epiAccessionsAreDirty = false;
}


/*
 * Add annotations to the tree
 */
void EpiAPI::addAnnotationsToTree(Tree* tree){
	if (!epiAnnotationsAreDirty) return;
	if (tree == nullptr) return;
	if (EPIDEMIOLOGY == nullptr) return;
	EPIDEMIOLOGY->addAnnotationsToTree(tree);
	epiAnnotationsAreDirty = false;
}




/*
 * List of annotations
 */
vector<string> EpiAPI::getAllAnnotations(){
	if (EPIDEMIOLOGY == nullptr) {
		vector<string> x;
		return x;
	}
	return EPIDEMIOLOGY->getAnnotations();


}


void EpiAPI::cleanup(){
	if (EPIDEMIOLOGY != nullptr){
		EPIDEMIOLOGY->cleanup();
		EPIDEMIOLOGY->clearAnnotationsFromTree(PhylogenyAPI::THE_TREE);
		delete EPIDEMIOLOGY;
		EPIDEMIOLOGY = nullptr;
	}
	if (timeline != nullptr){
		timeline->cleanup();
		delete timeline;
		timeline = nullptr;
	}
}

